### PR TITLE
docs(event_handler): add micro function examples

### LINKED ISSUE #902: More full examples of the micro function pattern
**What were you initially searching for in the docs?**

I am new to powertools and have been reading the docs to determine how to integrate it into my existing projects.  My project use the micro function pattern for API Gateway.  All of the API Gateway code examples and discussion are more relevant to the monolith pattern.  There is the discussion of monolith vs micro function, but no actual code examples for the micro function pattern.

For example, with the micro function pattern would one still use the ApiGatewayResolver when there is only one route?  Or would one use the @event_source decorator approach instead?

**Is this related to an existing part of the documentation? Please share a link**
https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/#considerations
https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/
https://awslabs.github.io/aws-lambda-powertools-python/latest/utilities/data_classes/#api-gateway-proxy

**Describe how we could make it clearer**
Please add more examples for the micro function pattern.  Please add more details about when you would use the ApiGatewayResolver vs the event_source decorator or if you would ever use both together.

**If you have a proposed update, please share it here**

### PR DESCRIPTION
**Pull request by TonySherman**
**Issue number:**
#902
## Summary
Adding an example of a micro function pattern  REST API

### Changes
No code changes. Documentation updates only
> Please provide a summary of what's being changed

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>
No
**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Hi @TonySherman! Can you enable our write access to this branch? We need to keep this PR in sync with our `develop` branch and we can't do that because you need to enable this option:

image

Thanks

**Comment by TonySherman** [comment]
> Hi @TonySherman! Can you enable our write access to this branch? We need to keep this PR in sync with our `develop` branch and we can't do that because you need to enable this option:
> 
> image
> 
> Thanks

It doesn't look like I have the ability to do this. Perhaps because it's a fork by my organization.

**Comment by leandrodamascena** [comment]
> > Hi @TonySherman! Can you enable our write access to this branch? We need to keep this PR in sync with our `develop` branch and we can't do that because you need to enable this option:
> > image
> > Thanks
> 
> It doesn't look like I have the ability to do this. Perhaps because it's a fork by my organization.

Aaa ok, don't worry about it. @sthulb and I will review this PR and provide appropriate feedback.

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (sthulb) [inline examples/event_handler_rest/src/micro_function_user_by_id_route.py]: the minor nitpick here is, this should be a 404 instead of 204.

**Review by sthulb — DISMISSED** [review]
These docs are good, thanks!

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (TonySherman) [reply examples/event_handler_rest/src/micro_function_user_by_id_route.py]: Updated. Thanks

**Comment by TonySherman** [comment]
Thanks for the review @sthulb! I made a change on the one comment you had. If this falls behind the develop branch again. feel free to ping me. I'll try to keep an eye on it too.

**Review by sthulb — DISMISSED** [review]

File: docs/core/event_handler/api_gateway.md
Comment (leandrodamascena) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
Each endpoint will be it's own Lambda function that is configured as a [Lambda integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/getting-started-with-lambda-integration.html){target="_blank"}. This allows you to set different configurations for each lambda (memory size, layers, etc.).
```

File: examples/event_handler_rest/sam/micro_function_template.yaml
Comment (leandrodamascena) [inline examples/event_handler_rest/sam/micro_function_template.yaml]:
```suggestion
    Runtime: python3.11
```

File: examples/event_handler_rest/sam/micro_function_template.yaml
Comment (leandrodamascena) [inline examples/event_handler_rest/sam/micro_function_template.yaml]:
```suggestion
            MemorySize: 128 # This Lambda Function can have a higher memory size set
```

It's better to keep 128 MB to avoid unexpected costs if someone wants to try this with many requests.

File: examples/event_handler_rest/src/micro_function_all_users_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_all_users_route.py]:
```suggestion
from dataclasses import dataclass
```

We should remove Pydantic to make this example easier to test and not force customers to bring external dependencies. We are not making validation and we can use dataclass instead. What do you think?

File: examples/event_handler_rest/src/micro_function_all_users_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_all_users_route.py]:
```suggestion
@dataclass
class User():
    user_id: str
    email: str
    active: bool
```

File: examples/event_handler_rest/src/micro_function_all_users_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_all_users_route.py]:
```suggestion
    all_users = [User(**user) for user in users]
    all_active_users = [user.__dict__ for user in all_users if user.active]
```

File: docs/core/event_handler/api_gateway.md
Comment (leandrodamascena) [inline docs/core/event_handler/api_gateway.md]: I really like this explanation, but do you think we can add something here about using the Lambda Layer to avoid code reuse in common/shared functions?

File: examples/event_handler_rest/src/micro_function_all_users_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_all_users_route.py]:
```suggestion
```

Remove this, please. Keep just 3 records and I think it's enough.

File: examples/event_handler_rest/src/micro_function_all_users_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_all_users_route.py]:
```suggestion
        "active": False,
```

File: examples/event_handler_rest/src/micro_function_all_users_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_all_users_route.py]:
```suggestion
def lambda_handler(event: dict, context: LambdaContext) -> dict:
```

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_user_by_id_route.py]:
```suggestion
def lambda_handler(event: dict, context: LambdaContext) -> dict:
```

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_user_by_id_route.py]:
```suggestion
@dataclass
class User():
    user_id: str
    email: str
    active: bool
```

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_user_by_id_route.py]:
```suggestion
```

Remove this, please. Keep just 3 records and I think it's enough.

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_user_by_id_route.py]:
```suggestion
from dataclasses import dataclass
```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Thanks for your review @sthulb!

Hi @TonySherman, I made some additional comments and it would be good to address them before we merge. This is a great opportunity to show our customers how to use a microfunction pattern, thank you very much for investing time in this.

File: examples/event_handler_rest/src/micro_function_user_by_id_route.py
Comment (leandrodamascena) [inline examples/event_handler_rest/src/micro_function_user_by_id_route.py]:
```suggestion
def get_user_by_id(user_id: str) -> Union[User, None]:
    for user_data in users:
        if user_data["user_id"] == user_id:
            return User(
                user_id=str(user_data["user_id"]),
                email=str(user_data["email"]),
                active=bool(user_data["active"]),
            )
            
    return None
```

Hey, total mistake on my part here! Mypy is complaining about types, so we have the latest changes :)

**Review by leandrodamascena — CHANGES_REQUESTED** [review]

**Comment by TonySherman** [comment]
Thanks for the suggestions @leandrodamascena! Let me know if there's any other changes you'd like.

**Review by leandrodamascena — APPROVED** [review]

**Comment by leandrodamascena** [comment]
Hey @TonySherman! Can you please update this branch? We will merge and include in this release!

Thanks a lot again, this is an important contribution.

**Pull request merged**
